core/eolearn/core/eodata_io.py

```diff
@@ -281,20 +303,69 @@ def _save(self, filesystem, data, path, file_format, compress_level=0):
             with gzip.GzipFile(fileobj=file_handle, compresslevel=compress_level, mode='wb') as gzip_file_handle:
                 self._write_to_file(data, gzip_file_handle, file_format)
 
-    @staticmethod
-    def _write_to_file(data, file, file_format):
+    def _write_to_file(self, data, file, file_format):
         """ Writes to a file
         """
         if file_format is FileFormat.NPY:
-            np.save(file, data)
-        elif file_format is FileFormat.PICKLE:
-            pickle.dump(data, file)
+            return np.save(file, data)
+
+        if file_format is FileFormat.GPKG:
+            return data.to_file(file, driver="GPKG", encoding="utf-8", layer=fs.path.basename(self.path))
+
+        if file_format in [FileFormat.JSON, FileFormat.GEOJSON]:
+            if self.feature_type is FeatureType.BBOX:
+                data = data.geojson
 
-    @staticmethod
-    def _decode(file, path):
+            if self.feature_type is FeatureType.TIMESTAMP:
+                data = [timestamp.isoformat() for timestamp in data]
+
+            try:
+                json_data = json.dumps(data, indent=2)
+            except TypeError as exception:
+                raise TypeError(
+                    f'Failed to serialize {self.feature_type} into a JSON file. Make sure that this feature type '
+                    f'contains only basic Python types before attempting to serialize it.'
+                ) from exception
+
+            return file.write(json_data.encode())
+
+        raise ValueError(f'Unsupported file format {file_format} for feature type {self.feature_type}.')
+
+    def _decode(self, file, path):
         """ Loads from a file and decodes content
         """
-        if FileFormat.PICKLE.extension() in path:
+        if path.endswith(FileFormat.NPY.extension()):
+            return np.load(file)
+
+        if path.endswith(FileFormat.GPKG.extension()):
+            dataframe = gpd.read_file(file)
+
+            if dataframe.crs is not None:
```

Comment:
Why this wrapping into `CRS` and back?

Reply:
It's because of the thing you also noticed yourself - if you save Geopackage and then load it back the CRS definition will not necessarily remain completely the same. It turns out that this happens not only for WGS84 but even for Popular Web Mercator and others. So by casting it to `sentinelhub.CRS` class and back we are correcting these definitions to the ones used in our code base and achieving more consistency in the process of saving and loading.